benchmark/http2/inheritance.js

```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+const common = require('../common.js');
+const PORT = common.PORT;
+
+const ORIGINAL = 'original';
+const SET_PROTOTYPE_OF = 'setPrototypeOf';
+const PARAMS = 'params';
+
+var bench = common.createBenchmark(main, {
+  streams: [100, 500, 1000, 2500, 5000, 10000],
+  version: [ORIGINAL, SET_PROTOTYPE_OF, PARAMS],
+  requests: [5000, 10000, 20000, 50000]
+}, { flags: ['--expose-http2', '--no-warnings'] });
+
+function main(conf) {
+  const http2 = require('http2');
+  const maxConcurrentStreams = Number(conf.streams);
+  const requests = Number(conf.requests);
+
+  class MyServerRequest extends http2.Http2ServerRequest {
+    getUserAgent() {
+      return this.headers['user-agent'] || 'unknown';
+    }
+  }
+
+  class MyServerResponse extends http2.Http2ServerResponse {
+    status(code, contentType) {
+      this.writeHead(code, { 'Content-Type': contentType });
+    }
+  }
+
+  let server;
+
+  if (conf.version === ORIGINAL) {
+    // Original prototype
+    server = http2.createServer((req, res) => {
+      MyServerResponse.prototype.status.call(res, 200, 'text/plain');
+      res.end(`User Agent: ${MyServerRequest.prototype.getUserAgent.call(req)}`
+      );
+    });
+  } else if (conf.version === SET_PROTOTYPE_OF) {
+    // Express style prototype extension for each request
+    server = http2.createServer((req, res) => {
+      Object.setPrototypeOf(req, MyServerRequest.prototype);
+      Object.setPrototypeOf(res, MyServerResponse.prototype);
```

Comment:
This closure is harder to optimize than the one for the PARAMS option because it has to access MyServerRequest & MyServerResponse. You would probably want to declare these outside of `main`.

Reply:
I tried to do something like Express https://github.com/expressjs/express/blob/a4bd4373b2c3b2521ee4c499cb8e90e98f78bfa5/lib/middleware/init.js#L35 to demonstrate the benefits of the new approach. I'm not sure we should merge the benchmark as well, it's more like a demonstration of this PR.